toolbox/errorfinder/main.go

```diff
@@ -0,0 +1,394 @@
+func getTokenFromWeb(config *oauth2.Config) *oauth2.Token {
+	authURL := config.AuthCodeURL("state-token", oauth2.AccessTypeOffline)
+	fmt.Printf("Go to the following link in your browser then type the "+
+		"authorization code: \n%v\n", authURL)
+
+	var authCode string
+	if _, err := fmt.Scan(&authCode); err != nil {
+		log.Fatalf("Unable to read authorization code: %v", err)
+	}
+
+	tok, err := config.Exchange(context.TODO(), authCode)
+	if err != nil {
+		log.Fatalf("Unable to retrieve token from web: %v", err)
+	}
+	return tok
+}
+
+// Retrieves a token from a local file.
+func tokenFromFile(file string) (*oauth2.Token, error) {
+	f, err := os.Open(file)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	tok := &oauth2.Token{}
+	err = json.NewDecoder(f).Decode(tok)
+	return tok, err
+}
+
+// Saves a token to a file path.
+func saveToken(path string, token *oauth2.Token) {
+	fmt.Printf("Saving credential file to: %s\n", path)
+	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		log.Fatalf("Unable to cache oauth token: %v", err)
+	}
+	defer f.Close()
+	json.NewEncoder(f).Encode(token)
+}
+
+func readSpreadSheet(credentialsPath string, spreadsheetID string, readRange string) []string {
+	b, err := ioutil.ReadFile(credentialsPath)
+	if err != nil {
+		log.Fatalf("Unable to read client secret file: %v", err)
+	}
+
+	// If modifying these scopes, delete your previously saved token.json.
+	config, err := google.ConfigFromJSON(b, "https://www.googleapis.com/auth/spreadsheets")
+	if err != nil {
+		log.Fatalf("Unable to parse client secret file to config: %v", err)
+	}
+	client := getClient(config)
+
+	srv, err := sheets.New(client)
+	if err != nil {
+		log.Fatalf("Unable to retrieve Sheets client: %v", err)
+	}
+
+	resp, err := srv.Spreadsheets.Values.Get(spreadsheetID, readRange).Do()
+	if err != nil {
+		log.Fatalf("Unable to retrieve data from sheet: %v", err)
+	}
+
+	if len(resp.Values) == 0 {
+		fmt.Println("No data found.")
+		return nil
+	}
+	filePaths := []string{}
+
+	for _, row := range resp.Values {
+		filePath := row[0]
+		filePathString, err := filePath.(string)
+		if err {
+			filePaths = append(filePaths, filePathString)
+		}
+	}
+
+	return filePaths
+}
+
+func NewErrorFinder(client *storage.Client, bucketName string) (*ErrorFinder, error) {
+	return &ErrorFinder{
+		client:     client,
+		bucketName: bucketName,
+	}, nil
+}
+
+func (f *ErrorFinder) query(ctx context.Context, prefix string) (string, error) {
+	client := f.client
+	bucket := client.Bucket(f.bucketName)
+	buildFile := bucket.Object(prefix + "build-log.txt")
+
+	rc, err := buildFile.NewReader(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer rc.Close()
+
+	data, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func (f *ErrorFinder) getContent(ctx context.Context, filePaths []string) (map[string]string, map[string][]string, map[string][]string, error) {
```

Comment:
Why do we need all of this manual parsing when the tests already output a well formatted junit file?

Reply:
I'm reading the build-log.txt files because I'm interested in errors that occur when running the commands that can be both fatal and not fatal that comes from the build and run commands but not from tests.